hyper/http20/connection.py

```diff
@@ -436,6 +442,17 @@ def _update_settings(self, frame):
 
             self._settings[SettingsFrame.INITIAL_WINDOW_SIZE] = newsize
 
+        if SettingsFrame.SETTINGS_MAX_FRAME_SIZE in frame.settings:
+            new_size = frame.settings[SettingsFrame.SETTINGS_MAX_FRAME_SIZE]
+            if new_size > FRAME_MAX_LEN and new_size < FRAME_MAX_ALLOWED_LEN:
+                self._settings[SettingsFrame.SETTINGS_MAX_FRAME_SIZE] = new_size
+            else:
+                log.warning(
+                    "Frame size %d is outside of allowed range",
+                    new_size)
+                # Tear the connection down with error code PROTOCOL_ERROR
+                self.close(1)
+
```

Comment:
I think, rather than tie ourselves in knots trying to handle teardown mid-connection, we should throw exceptions whenever we have to tear a connection down ourselves. That means, still call `self.close()`, but then immediately throw an exception.

While we're here, it might be good to define an exception that we throw whenever we encounter a HTTP/2 connection error.